File: python/day-10/a.py
from typing import List


def main():
    lines: List[str] = open('in').read().split()
    illegal_chars: List[str] = get_illegal_chars(lines)
    score: int = get_score(illegal_chars)
    print(score)    


def get_illegal_chars(lines: List[str]):
    return [illegal_char for line in lines if (illegal_char := find_illegal_char(line)) is not None]


def find_illegal_char(line: str):
    open_chunk_chars = ['(', '[', '{', '<']
    close_chunk_pairs = {
        ')': '(',
        ']': '[',
        '}': '{',
        '>': '<'
    }

    stack: List[str] = []
    for c in line:
        if c in open_chunk_chars:
            stack.append(c)
        elif stack.pop() != close_chunk_pairs[c]:
            return c
            
    return None


def get_score(chars: List[str]):
    char_scores = {
        ')': 3,
        ']': 57,
        '}': 1197,
        '>': 25137
    }

    return sum(char_scores[c] for c in chars)


main()

File: rust/day-10/a.rs
use std::fs;
use std::error::Error;
use std::collections::HashMap;


fn main() -> Result<(), Box<dyn Error>>{
    let inp = fs::read_to_string("in")?;
    let lines: Vec<&str> = inp
        .lines()
        .collect();

    let illegal_chars: Vec<char> = get_illegal_chars(lines);
    let score: usize = get_score(illegal_chars);

    println!("{}", score);
    Ok(())
}


fn get_illegal_chars(lines: Vec<&str>) -> Vec<char> {
    lines
        .iter()
        .filter_map(|line| find_illegal_char(line))
        .collect()
}


fn find_illegal_char(line: &str) -> Option<char> {
    let open_chunk_chars = vec!['(', '[', '{', '<'];
    // For some reason this gives an error
    // let close_chunk_pairs: HashMap<_, _> = [
    //     (')', '('),
    //     (']', '['),
    //     ('}', '{'),
    //     ('>', '<')
    // ].into();
    let mut close_chunk_pairs = HashMap::new();
    close_chunk_pairs.insert(')', '(');
    close_chunk_pairs.insert(']', '[');
    close_chunk_pairs.insert('}', '{');
    close_chunk_pairs.insert('>', '<');

    let mut stack: Vec<char> = Vec::new();
    for c in line.chars() {
        if open_chunk_chars.contains(&c) {
            stack.push(c);
        } else if stack.pop().unwrap() != close_chunk_pairs[&c] {
            return Some(c);
        }
    }
            
    None
}


fn get_score(chars: Vec<char>) -> usize {
    // For some reason this also gives an eror
    // let char_scores = HashMap::from([
    //     (')', 3),
    //     (']', 57),
    //     ('}', 1197),
    //     ('>', 25137)
    // ]);

    let mut char_scores = HashMap::new();
    char_scores.insert(')', 3);
    char_scores.insert(']', 57);
    char_scores.insert('}', 1197);
    char_scores.insert('>', 25137);

    chars
        .iter()
        .map(|c| char_scores[c])
        .sum()
}